### ISSUE TITLE
Concrete Syntax in tree

### ISSUE THREAD

**Issue by getify** [estree/estree#41]
I'd like to open up the discussion of what our various options are for preserving so called "Concrete Syntax" elements in the parse tree format.

## What are Concrete Syntax elements?

A simplified definition of "concrete syntax" is any stuff that could appear in a source code program which, when it's parsed, would be otherwise discarded information that an AST (as it stands now) does not represent.

That specifically means that things which are reliably inferable from an AST's nodes are *not* concrete syntax. For example:

```js
var a = 2,
    b = (a + 2) * 3;
```

Here, the `( )` around `a + 2` is not represented in the AST, because the structure of the tree, combined with operator precedence rules, absolutely implies that it must exist, and moreover `a + 2 * 3` would have been a different tree structure. Thus, the `( )` in the original program can be reconstructed from the AST reliably, so it need not be stored separately. **It is not concrete syntax**.

However:

```js
var a = 2,
    b = a + (2 * 3);
```

This program includes a pair of `( )` that would already be implied by the tree structure and operator precedence, and thus would also not be stored.

But, critically, they also would not be re-generated when the tree was reconstituted. Why? Because it's impossible to know from the tree alone if the `( )` was really there, or just implied. And code generation takes the conservative path and doesn't make up `( )` where it's not sure they were there, so it leaves them out.

Herein we see that the original `( )` are not abstract syntax, but concrete, and to preserve them is going to require some other solution.

Examples of concrete syntax:

* extraneous whitespace (the non-significant kind)
* comments
* extraneous `( )` (used primarily for readability more than functionality)
* ... other?

### The Rabbit Hole

Just how deep does this rabbit hole go? In the following snippet, every single `/*x*/` comment represents a location where whitespace and/or comments can optionally appear as concrete syntax, in addition of course to the fact that some of those locations require significant whitespace, such as `/*4*/`:

```js
/*1*/class /*2*/ Foo /*3*/ extends /*4*/ Bar /*5*/ {
   /*6*/constructor /*7*/ (/*8*/x /*9*/ = /*10*/ "hello" /*11*/) /*12*/ {
      // ..
   }

   /*13*/ static /*14*/ bam /*15*/ ( .. ) {
      // ..
   }
}

var /*16*/{ /*17*/ x /*18*/: /*19*/ { /*20*/ y /*21*/: /*22*/ z /*23*/ = /*24*/ 2 /*25*/ } } =
   /*26*/new /*27*/ Foo /*28*/ ( (/*29*/) /*30*/ => /*31*/ { .. } /*32*/ ) /*33*/;
```

Your imagination can probably take it from there. There's a whole slew of complex ES6 syntax forms which implies a deep rabbit hole of nooks and crannies where we need to be able to preserve information (in some way) that our normal approach to AST doesn't currently preserve.

Consider the tree structure for the arrow function expression, for example... how and where could we represent `/*29*/` and `/*30*/`? `/*31*/` and `/*32*/` are a little clearer.

## Why Concrete Syntax?

Why would we want to preserve all these pieces of concrete syntax? There's several use-cases:

* Any tool which performs localized transformations on a source code file, which doesn't want to change everything, but only a targeted specific thing. For example, a tool that does nothing but rewrite all variable names to uppercase versions (for whatever silly reason).

   The goal of this tool is not to affect anything else about the program, such as formatting and comments, as those might still be important to the author of the file.
* Fully configurable automated code formatting: not just rule based, like "always put a space between an `( )` on a call" or stuff like that, but more fine-grained rules, where a tool might parse a source program and produce a tree structure with very specific information in it about how the resultant code should be re-generated.

   For example, it may automatically insert comments for each parameter in a function declaration with some sort of annotations about how and where the param is lexically used, etc. Or a code-style painter may "repaint" a file with spaces for indentation vs tabs, or may insert spaces for alignment and indentation with tabs, etc etc.

I could go on speculating, but I'll just leave it that there are definitely cases for tools which want to be able to preserve concrete syntax wherever it appears. Since concrete syntax by definition cannot be inferred, the parser and data structures that come out of it must be able to do so. Moreover, this information must be something a code generator can receive and use.

## How?

Here's the part where all the bikeshedding is going to happen.

To date, conversations around this topic have happened many times that I've been privy to, and there's never been any kind of consensus on how to approach solving it. I have my ideas, but I'm only going to suggest them as a possible starting point proposal, not that it has to be this way.

### CST-as-AST Proposal

I believe we should have one unified tree structure, which has optional -- what I call "extras" -- annotations (and in a few limited cases, nodes) in it which represent the necessary hooks for preserving these concrete syntax elements. In other words, I propose that there be no difference between an AST and a CST (concrete syntax tree), other than the absence or presence of CS elements in the tree.

Any tool which currently produces ASTs is a tool that's producing CSTs by default, but which just happens to not actually be keeping any of the CS elements yet. These tools can start keeping the CS elements, but still have the same style of tree they're producing, just with extra info in them.

Any tool which consumes ASTs is a tool that's already consuming CSTs by default, but which is just ignoring any CS elements which may be present. These tools can start using the CS elements they find.

It turns out that *most* of the places where we need to preserve CS elements can be added as additional properties (again, I call it "extras", with extra sub-names like "before", "inside", and "after" for positioning), which means that there would be **zero impact** to the existing tools that use such a tree format.

Tree producing tools (parsers, transpilers, etc) could just not produce these annotations, but things still work fine. Tree consuming tools continue to consume the trees as they currently do, and just ignore the these extras as they currently do.

I believe this has the most minimal impact to the existing tool ecosystem, and thus the easiest path to wider adoption by more tools.

#### Downside

There will be some minor places where the node structure will have to be slightly different to accommodate some of the trickier cases of CS positioning.

For example, anonymous function expressions that have an `id` of `null` means that we don't have an object value in that node to attach any `extras` annotations to in the `function/*1*/()` position.

If we could represent an anonymous name entry instead of `id: null` as an object like `id: { extras: .. }`, this will mean we have a hook to annotate those extras.

This *does* represent a slight breaking change to the format, but it's not a major sweeping new tree format, and will require on the whole just a small bit of extra handling care. These necessary node structure breaking changes are minor and very few for the pre-ES6 tree structure (aka SpiderMonkey).

The new ES6 forms definitely add more places where we should consider tree structure from the beginning which are amenable to attaching these annotations. Since there's not already an established standardized ES6 tree format, I think it's not too late for us to consider these concerns as we specify the ES6 node forms.

**Comment by michaelficarra** [comment]
> ... other?

Off the top of my head:

* escape sequences
* string delimiters
* number formatting
* high precision or very large (magnitude) numbers
* almost all semicolons
* others depending on AST structure

**Comment by getify** [comment]
@michaelficarra great points, knew I was forgetting some of them. :)

**Comment by michaelficarra** [comment]
> Any tool which performs localized transformations on a source code file, which doesn't want to change everything

I disagree that this would require full concrete syntax information. Localised replacements can be done with source position information alone. But I still agree that concrete syntax information has its uses.

> I believe we should have one unified tree structure

I completely agree. This is the evolutionary route forward.

> For example, anonymous function expressions that have an `id` of `null` means that we don't have an object value in that node to attach any `extras` annotations to in the `function/*1*/()` position.

Exactly. This is one of the many reasons I have argued in the past that each piece of concrete syntax information should have its own named property on each node. For example, each position whitespace/comments are allowed would have a named field in the CST. Additionally, each of the concrete syntax elements I mentioned in my comment above need to be represented on the appropriate node. This is incompatible with a simple string for "before/after/in-between" concrete syntax.

Finally, are we sure we want to have this discussion here? It seems a bit out of scope for this project. Also, readers should check out previous CST discussions/proposals:

* https://github.com/estools/escodegen/wiki/CST-Proposals
* https://github.com/getify/concrete-syntax-tree

**Comment by sebmck** [comment]
@michaelficarra 

> Finally, are we sure we want to have this discussion here? It seems a bit out of scope for this project. 

Completely agreed. We haven't even agreed on the entire ES6 AST spec and there's still a lot of work to do so getting sidetracked into CST discussion so early is only going to be harmful IMO.

**Comment by mikesherov** [comment]
I agree we should defer discussion of CST until we get past thorny bits of ES6, however, we need to develop the ES6 bits with the CST in mind as a future concern. That is to say, when considering ES6 features, we should consider whether the specced feature is harmful to future CST implementation.

**Comment by mikesherov** [comment]
There is also a ton of prior art here in JSCS, ESLint, ESFormatter, and friends to how a CST may work. In terms of a specific proposal, I have ideas based off what both JSCS and esformatter have arrived at.

The basic concept is to formalize the token list, and leave the AST as is, except have a formal link between nodes in the tree and the position of the first and last tokens of that mode in the token list.

The AST portion needs to then formalize what constitutes the tokens of a node (which includes question like "what about grouping parens and semicolons). Each node in AST would have a `firstToken` and lastToken property that points to the tokens in the token list. 

Once you have a link between the tree and the tokens, you augment the token list to be a doubly linked list (for ease of traversal or insertion) that also includes comment tokens, introduces an EOF token, and adds a `whitespaceBefore` property to each token representing the non comment whitespace between tokens.

Rerendering the source file is as simple as iterating over all tokens and concatenating whitespaceBefore with value of each token.

Augmentation of nodes remains trivial because you can just do a linked list replacement on the replaced nodes first and last token.

This separation allows abstract consumers to not really think about concrete, and vice versa.

**Comment by mikesherov** [comment]
Note that the approach above is essentially how both popular AST based formatters JSCS and esformatter augment the existing AST to generalize their approach.

**Comment by millermedeiros** [comment]
I would really like a format that makes it easier to add/remove nodes and tokens, augmenting the AST won't make it easy to manipulate.. each node type will require an unique logic. reconstructing the code will be hard (can't do a simple string concatenation). 

The linked list (or array of tokens) have some advantages:

 - easy to [loop/search inside a range](https://github.com/millermedeiros/rocambole-token/blob/master/find.js)
 - easy to [add](https://github.com/millermedeiros/rocambole-token/blob/master/insert.js) and [remove](https://github.com/millermedeiros/rocambole-token/blob/master/remove.js) tokens
 - easy to replace the token value (eg. convert [double quotes into single quotes](https://github.com/millermedeiros/esformatter-quotes/)..)
 - very easy to reconstruct the code (just need to [concat all the `token.value`](https://github.com/millermedeiros/rocambole/blob/32595d678e3df52ca2831056fc450f46529d4df8/rocambole.js#L104-L113))

Drawbacks of liked list:

 - main drawback is that you don't have a link between the tokens and the nodes, so:
   - if you add a new element it won't be part of the AST (which might negate some of your logic and/or make things way more complex)
   - adding new tokens at the *edges* of the node might cause some undesired side effects (reference to `startToken` and `endToken` on the AST will point to wrong tokens..)
   - hard/expensive to detect all the nodes that contains that token (all tokens are part of `Program`)
 - hard to find the matching `()`, `{}`, `[]` (need some sort of state machine that counts how many opening/closing parenthesis in between..)
 - can't be serialized as JSON given the circular references

In my opinion the best format for code manipulation would be structured like a tree but would be more similar to nested arrays than to nested objects, it would have the following characteristics:

 - easy to insert new nodes and tokens before/after/**around**/inside any node
 - easy to remove/replace nodes and tokens
 - easy to convert back into a string
 - easy to find matching `()`, `{}` and `[]`
 - straightforward logic to loop through all the tokens inside a node/program (changes can be made without full context of the node/token)
 - can be serialized as JSON
    -  means that we would need helper libs or maybe even *enhance* the CST with extra properties (eg. reference to `parent` node) to be able to traverse it easily (imagine you need to get a reference to parent node and/or next/previous token)

The CST wouldn't need any range/loc info, specially since these get **outdated after each manipulation**.

**Comment by getify** [comment]
> main drawback is

I'd say the main drawback to having anything more than just one complete data structure is that you make it all but impossible/impractical for different tools in the flow of parsing->transformation->codegen to compose through typical methods, such as command line I/O piping, etc.

The secondary drawback, as you allude to, is that you've doubled the work to do any transformations, since you have to update both if there's any hope of keeping a sane flow of transformations between tools.

I cannot even fathom it being a good idea to have multiple structures. To me that is a complete show stopper design.

**Comment by millermedeiros** [comment]
> impossible/impractical for different tools in the flow of parsing->transformation->codegen to compose through typical methods, such as command line I/O piping, etc

not totally true. depending on the kind of manipulation you don't need to touch the AST structure - on [esformatter](http://github.com/millermedeiros/esformatter/) we just manipulate the tokens (mainly WhiteSpace, LineBreak, Indent and comments), so there is no need to update both structures. - that's why I decided to implement the linked list, it's the simplest structure that allow us to add/remove/search tokens, but it's less than ideal for complex manipulations (that's why I still want a real CST).

you always have the option to parse the code multiple times (command line I/O piping usually works over strings).. and that's really the path we are going to take on esformatter moving forward for plugins/tools that needs to change the code structure (see: https://github.com/millermedeiros/esformatter/issues/168) - if all tools receive a string as input and output a string it's easier to mix and match, they can use whatever format is better for the job.

there is no need for complex `codegen` when the data structure already contains all the tokens in the proper order (codegen becomes just a basic string concatenation) - `toString()` on rocambole is incredibly fast.

**Comment by getify** [comment]
you're only talking about one specific view of your use-case. i'm talking about making it more difficult in the general case for all potential tools in the ecosystem.

**Comment by millermedeiros** [comment]
> you're only talking about one specific view of your use-case. i'm talking about making it more difficult in the general case for all potential tools in the ecosystem.

yes, I'm looking through the eye of someone who needs to insert WhiteSpace/LineBreak around tokens, WhiteSpace/Indent/LineBreak inside elements, change between comma-first and comma-last, re-order variable declarations, convert between single variable declaration and multiple, normalize quotes around strings, [and so many other things that are currently handled by esformatter plugins](https://github.com/millermedeiros/esformatter/wiki/Plugins)... my point of view is that your proposal of shoehorning the AST is not going to make my work any simpler, in fact it would make it harder.. - maybe it would make your work simpler, I have no idea what are your use cases..

adding any extra property to the AST might cause undesired side effects; for instance esprima@2.1.0 added the `handler` property to every `TryStament` together with the `handlers` array.. which made *rocambole* loop over the same `CatchClause` twice, breaking the *esformatter* indentation logic (it was adding 2 indents instead of 1). even tho you currently think that keeping the same structure will be easier, it won't be backwards compatible and current tools won't magically work out of the box, there is always a lot of work to be done to adapt things and make sure we all play by the same rules.

**Comment by getify** [comment]
OK, so let's for now grant that no one solution is perfect for all use-cases. Whatever solution we pick will cause some use-cases to be relatively easier than today's ad hoc options, or relatively harder, and some use-cases will be flipped.

----

But what *is* true, and was my earlier point, is that if two CLI utilities want to cooperate, and the sharing (via I/O piping) is currently just tool A outputting a string on the stream (say the JSON of the tree) that tool B then consumes (not re-parses from scratch) this tree and makes more changes, and re-outputs another tree, and ... THAT process will be more complicated if each tool needs to share two separate structures.

If we take CLI I/O piping out of the concern, and think only about streams in node (think gulp), the same concern applies. And if we take stream-based I/O entirely off the table, every API method in every tool which currently takes one argument (the tree or whatever) now has its arity/complexity doubled to accept two data structures.

THAT is all I meant by suggesting that across the ecosystem, passing around two structures is twice as complicated as passing around one structure.

**Comment by millermedeiros** [comment]
totally agree, better to have a single structure that contains everything you need. less moving parts. I just don't think that adding extra properties to an AST is enough, it would still be hard to do some of the manipulations that I described on my previous comment. I'd rather have a new format that makes it easier to rearrange nodes while also making it simple to edit the tokens that compose the program.

**Comment by mikesherov** [comment]
> I'd rather have a new format that makes it easier to rearrange nodes while also making it simple to edit the tokens that compose the program.

You can have that. What you won't have though is collaboration with a bunch of other tools that need the existing tree structure.

Adding the tokens to the structure with ability to reference between tree and token list solves majority of use cases, is immediately back compat, works with structure and whitespace modification, and has the only downside of some theoretic difficulty in maintaining both, which is just not proven to be true.

If you'd like to continue having string output and have every tool in pipeline pay the parse penalty, that's also fine, but we're talking about passing around a single structure that is non lossy and only gets printed back to a string at the end.

Not paying the multi parse penalty is the key here.

**Comment by getify** [comment]
FWIW, to some of the earlier questions about whether this is a useful/necessary discussion to be having in scope of ESTree, I think this present line of discussion is precisely why some "standards body" (used in a very loose sense here, obviously) definitely needs to be taclking this.

To me it would be folly for some other external group to come along **later** and try to define a separate standard for how these token lists or annotations or whatever were somehow integrated into the tree we specify here. I think **we precisely** have to consider these issues and work through them, as the greater tooling ecosystem clearly needs something more than one tool's ad hoc solution can provide. The lack of such a coordinated broad effort is why this problem remains partially hacked around on a per-tool basis after years of efforts and complaints.

If ESTree doesn't standardize concrete syntax as part of the overall process here, then some other tree standard (entirely duplicative of ESTree) will have to do so, and **that** standard will have to be what folks like me hope most tools in the ecosystem adopt eventually.

IOW, if ESTree doesn't agree to the notion of standardizing concrete syntax (in some fashion, and at some point eventually) as part of the overall tree standard, then I think my next appropriate step will be to fork ESTree and continue that effort in parallel, and then hope to convince tools that the forked wider standard is what they consider, either now or sometime down the road. It'd be a lot less waste of effort to just define that as **in scope** for ESTree itself, and keep the effort centralized.

Anything short of that accomplishes precisely zero of what I've been on-and-off banging a drum about for the better part of two years.

**Comment by mikesherov** [comment]
@getify take it easy, man. No reason to get so worked up and go talking about a fork already! We're here trying to unite community and not divide.

No one said not to consider a CST, we're just talking about prioritizing the work here. First is finishing ES6, and then figure out how to proceed with CST. 

There is no need to discuss this in such an adversarial tone. The group here is rightfully asking whether CST is part of this discussion. That doesn't mean we won't tackle it, together.

**Comment by getify** [comment]
1. I deliberately didn't respond right away several days ago, so as to not just be "worked up". I am not worked up nor emotional about this. I just feel strongly, as this topic is something I've been trying to address for years now.

   I mentioned "fork" not as a threat but as a point about the proper scope, and specifically as something I would like to do everything in my power to avoid needing to do. A fork would be worst-case outcome IMO. For the same reason, years ago when I needed this concrete syntax stuff, I didn't just fork a parser and do all my own ad hoc work. My project *esre* has just been sitting waiting hoping for a solution. :/
2. "No one said not to consider a CST". well...

   > Finally, are we sure we want to have this discussion here? It seems a bit out of scope for this project. 

      
   > Completely agreed.

   I'm only reacting to those sentiments. Nothing more.
3. "prioritizing the work here" I don't mind prioritizing. I didn't start this thread demanding that we tackle all of CST now. That having been said, I strongly agree with...

   > we need to develop the ES6 bits with the CST in mind as a future concern

   And so I was merely saying, it only makes sense that we have a general idea of the overall direction (ie, that we don't consider a completely separate data structure) while we're thinking about the other tree node definitions.

   Specifically, I would very much like to avoid the "it's too late to do that" arguments that have actually specifically been used to thwart some of my earlier efforts.

Sorry for the alarming tone my previous message came across as. That's not actually my intent at all.

**Comment by mikesherov** [comment]
Thanks for clearing it up!

**Comment by nzakas** [comment]
Can we agree on a goal for this thread? Right now it doesn't seem like the conversation is heading towards any sort of conclusion. So far it seems people are agreed that:

1. A CST is desirable
1. It would be nice to have a CST built on top of ESTree rather than something separate

What are the next steps here?

**Comment by getify** [comment]
My goal was purely to get consensus on:

1. That concrete syntax is an important problem to solve
2. That ESTree should handle it in some form or fashion during the process of defining this new (aka updated) cross-tool standard.
3. That in some fashion this be a single data structure instead of separate meta-data.

The bikeshedding of how to do it, or even precisely when, is not in scope for this thread IMO. I probably shouldn't have even included my own proposal as that invites such.

**Comment by nzakas** [comment]
Based on those goals, have we concluded this thread at this point? If so, what would the next steps be?

**Comment by michaelficarra** [comment]
I can get behind 1 and 3 but disagree on 2. This project, as I understand it, is simply tasked with unifying and documenting the SpiderMonkey-based ASTs produced and consumed by popular tooling. Not making it better or fixing it, and especially not expanding its scope. As long as this isn't getting in the way, it should be a separate effort.

**Comment by mikesherov** [comment]
@michaelficarra I disagree with that. There's no reason this group can't and shouldn't tackle CST once we have ironed out the inconsistencies and ES6. It is precisely the major parsers and the major parser consumers who should be defining the CST IMO. This would have benefits for esformatter, JSCS, ESLint, Recast, Babel, etc... all of who consume, not surprisingly, Esprima, Acorn, or Spidermonkey.

Saying the goal of this group is not to make the SM AST better is a non-starter conversation for me. As long as the changes are additive and BC is preserved, we should not limit ourselves to being done once the wrinkles are ironed out.

**Comment by sebmck** [comment]
@michaelficarra The SpiderMonkey AST isn't going away (in fact some people such as me prefer it). Try not to let your bias interfere too much with the projects direction.

**Comment by mikesherov** [comment]
@RReverser @sebmck thoughts? Once we have concluded ironing out inconsistencies and ES6/7, are you interested in pursuing defining the CST portion of the ESTree?

**Comment by sebmck** [comment]
@mikesherov Sure I don't see why not. Would make it much easier for linters and easier for transformers that want to retain input whitespace etc.

**Comment by millermedeiros** [comment]
My next step is try to convince that keeping the same structure as the AST
is not a good idea. If that doesn't work I'll let you guys move forward and
keep doing my own thing.. Maybe that will motivate me to create a competing
*standard* - only reason why I haven't done it already is because rocambole
was good enough for basic token manipulation (like we need for esformatter)
and time is a very limited resource...

**Comment by mikesherov** [comment]
@millermedeiros if the conversation you want to have is changing the AST structure as you already described, then yes, you should create a competing standard. Anything that is non additive, or that breaks BC at all is not currently on the table.

**Comment by sebmck** [comment]
@millermedeiros 

> My next step is try to convince that keeping the same structure as the AST is not a good idea. If that doesn't work I'll let you guys move forward and keep doing my own thing..

Feel free to suggest specific changes or raise specific pain points. Discussion about changing the entire AST wont be taken seriously.

**Comment by millermedeiros** [comment]
Already gave the feedback about pain points and the kind of transformations
I want to do to the source code (which aren't easy with an AST even if you
add properties to hold whitespaces, comments, line breaks and parenthesis).
So the only option I have is really to work on my own thing while I wait
you guys realize it was a bad idea... Cheers

**Comment by mikesherov** [comment]
@nzakas to address your point, I think we need to lay down some basic principles for a CST so we can close this thread. I'll submit a PR that add a page that serves as an informative note about the CST that closes this issue.

1. Must be an additive change.
1. Must promote the concept of mutability, that is, augmenting the CST should not invalidate non local nodes.
1. Must be JSON serializable, so that tools can eventually output a parsed JSON string as input to other tools (to avoid the double parse penalty)
1. Must promote traversability. Should be able to get to the "next" or "prev" set of tokens from any node.
1. Must promote printability. Should be easy to generate the src from the tree.
1. Must be lossless. Seems obvious, but worth mentioning.

OK?

**Comment by getify** [comment]
+1

**Comment by nzakas** [comment]
Sounds good. I really just wanted to move on from the bike shedding to something more goal oriented. This sounds like a good way to do that.

**Comment by RReverser** [comment]
> Once we have concluded ironing out inconsistencies and ES6/7, are you interested in pursuing defining the CST portion of the ESTree?

Of course (I suppose everyone here is interested in that).

**Comment by jzaefferer** [comment]
@mikesherov that list seems to exclude many architectural approaches prematurely. This...

> Should be able to get to the "next" or "prev" set of tokens from any node.

...even prescribes a specific approach, one that Miller has argued against for a CST. I find that quite compelling, considering that he wrote rocambole, which implements exactly what you're suggesting. If the author of such a tool talks about the architecture shortcomings of said tool and suggests discussing alternatives, I'd hope for such a discussion to be welcomed, not shut down immediately.

> and has the only downside of some theoretic difficulty in maintaining both, which is just not proven to be true.

Then give Miller a chance to prove those difficulties. Considering the rather open timeline for CST proposals, that is unlikely to hurt progress, since he could do that while the ESTree team continues working on ES6 details.

**Comment by sebmck** [comment]
@jzaefferer Yes it's true that it may not be ideal for a CST but ESTree is an AST first and foremost and there are lot of different types of consumers. Having a balance between all of our needs is hard and we sometimes need to make compromises. We don't live in a perfect world where everyone gets what they want. Yes, a CST spec designed from the start may be suitable for certain tools but it'd likely be completely useful for a category of others.

Backwards compatibility will not be broken. ESTree was created as a collaborative effort between all the consumers so we have one spec that increases interoperability between the various tools. It's not really a good idea to lecture us on the shortcomings of the current representation or even a proposed one since there are a lot of people involved in this conversation that have a lot of experience doing this sort of stuff. Please don't be under the impression that we're disregarding input into this process, we're not. This conversation is just premature until everything else is out of the way so everyone's effort can be focussed on an interoperable and agreeable CST spec for all.

It doesn't make much sense to complain about a process that hasn't even happened yet. As @nzakas said can we please refrain from bikeshedding until the appropriate time since this conversation is very unproductive on it's own. Some patience and composure is more than appreciated in a standards process like this, especially one with so many interested parties involved.

**Comment by millermedeiros** [comment]
Just to make things clear:

My suggestion is to have 2 discrete formats.. one for the ABSTRACT code structure like we have right now (used by transpilers, type inference, code completion, etc..) and another format for the CONCRETE code structure (used by linters, code formatters, any code transformation) - `AST !== CST`

They don't need to be linked in any way, we could have parsers that only generate the AST, and parsers that only generate the CST. The same way that we would have tools that only work over the CST and tools that only handle AST.

I strongly believe that adding more properties to the AST is a bad idea. Maybe one day I'll have enough time/patience to write code that shows the kinds of operations that are [too hard to do with an AST](https://github.com/estree/estree/issues/41#issuecomment-78372421) and how a different format could look like and how [it would simplify things](https://github.com/estree/estree/issues/41#issuecomment-78285525).

Some flaws of the CST principles listed above:

 - Linked lists can't be serialized into JSON (how would you describe cross/circular references?) so the whole idea of "next" and "prev" tokens doesn't work out.
 - AST is not easily printable, see [escodegen](https://github.com/estools/escodegen) source code for an example. You need to infer a lot of things based on the tree structure since it doesn't contain all the tokens (if it doesn't contain all the parenthesis, colons, braces it is not a **concrete** syntax tree). Adding more properties to the tree will increase the codegen complexity even more...
 - if the AST contains range/loc info, it is not easy/cheap to mutate it (you need to update the loc info on all the nodes of the tree after each change).
 - additive changes are not always backwards compatible since tools might rely on the current tree structure to work.

If you all consider that a new format is not a valid solution, then this will eventually fail because the *correct* answer to this problem is definitely elsewhere.

**Comment by sebmck** [comment]
@millermedeiros 

> Linked lists can't be serialized into JSON (how would you describe cross/circular references?) so the whole idea of "next" and "prev" tokens doesn't work out.

Linked lists was never implied, only that you can get them easily.

> AST is not easily printable

Sure it is. [This](https://github.com/babel/babel/blob/master/src/babel/generation/generators/statements.js) looks pretty simple and easy to me :smile:.

> additive changes are not always backwards compatible since tools might rely on the current tree structure to work.


Which is why it's designed in such a way that it's not backwards incompatible.

> If you all consider that a new format is not a valid solution, then this will eventually fail because the correct answer to this problem is definitely elsewhere.

Dude, come on. There's no **easy** solution and pessimism wont gets us anywhere.

---

As I said in my previous comment, these are changes that would benefit **you**. ESTree has to take more into consideration than just you and your needs, an entire ecosystem relies on this stuff and interoperability is a big deal which is why we want to collaborate and compromise on something that works well for most rather than a few.

**Comment by RReverser** [comment]
> They don't need to be linked in any way

Sorry, but that sounds like a terrible idea. It might work for tools that need only CST like formatters, but it won't work nicely for tools that **do** need to work with both AST and CST as they will need to support two different formats with two different codebases to work with them - it might be `ESLint` that checks both structural correctness and code formatting, it might be `Tern` that uses both type inference and JSDoc comments, it might be `recast` that needs to transform code while preserving original formatting, and much, much more tools that benefit from tight coupling of structure and formatting information.

The fact that separate unlinked CST format works for one use-case doesn't mean it will work for others.

**Comment by millermedeiros** [comment]
> Sure it is. This looks pretty simple and easy to me :smile:.

compared with the 8 lines of code that I need on the [rocambole `toString()` method](https://github.com/millermedeiros/rocambole/blob/93be52fca355131db2916417918507d0f9fc39a7/rocambole.js#L104-L113), it's pretty complex, specially if you add the original format to the mix.. there are many other files inside the "babel/generation/" folder..

 > Sorry, but that sounds like a terrible idea. It might work for tools that need only CST like formatters, but it won't work nicely for tools that do need to work with both AST and CST

The CST can have the full structure (you know it's an Array, FunctionExpression, etc..) but it's built in a way that keeps ALL the tokens in the correct order.

 > The fact that separate unlinked CST format works for one use-case doesn't mean it will work for others.

Same can be said about the monkey-patched AST... No single format will cover all the use cases in the *optimal* way.

**Comment by sebmck** [comment]
> compared with the 8 lines of code that I need on the rocambole toString() method, it's pretty complex

Add some dynamically created nodes into the mix and you wont be saying the same thing.

> specially if you add the original format to the mix.. 

Yes and it's not possible without a CST or some screwing around with tokens. This is why whatever CST is designed needs to work well for **all** tools.

> there are many other files inside the "babel/generation/" folder..

Yes, one function for each node type.

**Comment by RReverser** [comment]
> The CST can have the full structure (you know it's an Array, FunctionExpression, etc..) but it's built in a way that keeps ALL the tokens in the correct order.

Are you suggesting to create completely different format that reflects all the same things but in new, non-backward-compatible way? Then it's not an option (and actually not even in scope of) for this project. Backward compatibility is one of the key stones of ESTree evolvement as it's a living standard for huge tools ecosystem that relies on SpiderMonkey-based format.

**Comment by mikesherov** [comment]
> ...even prescribes a specific approach, one that Miller has argued against for a CST. I find that quite compelling, considering that he wrote rocambole, which implements exactly what you're suggesting. 

@jzaefferer every one here has written their own post processor of the AST. The fact that rocambole does it a specific way isn't special to me. 

> If the author of such a tool talks about the architecture shortcomings of said tool and suggests discussing alternatives, I'd hope for such a discussion to be welcomed, not shut down immediately.

The discussion of things that are not additive to the AST are out of scope. @millermedeiros has stately explicitly he wants an alternative format.

> Then give Miller a chance to prove those difficulties. Considering the rather open timeline for CST proposals, that is unlikely to hurt progress, since he could do that while the ESTree team continues working on ES6 details.

It's made abundantly clear from the existing discussion that we're not on the same page with the GOALS, let alone the implementation.

**Comment by millermedeiros** [comment]
> Add some dynamically created nodes into the mix and you wont be saying the same thing.

It's a linked list.. you can add new nodes/tokens anywhere you want to; You just push the new tokens into the linked list and they will be printed - can even be a single token with an invalid type containing the full string of the node you want to insert. But as I said on my [first comment](https://github.com/estree/estree/issues/41#issuecomment-78285525), it's less than optimal.

 > Are you suggesting to create completely different format that reflects all the same things but in new, non-backward-compatible way?

Yes, I'm suggesting a new format that would be used by formatters, linters and any other tool that needs to modify and/or analyze the source code taking into consideration the original indentation, line breaks, parenthesis, colons, comments and any other type of information that isn't represented by an AST.

ASTs are extremely valid for transpilers and static analysis, but not so much when it comes to lossless code transformation.

**Comment by sebmck** [comment]
> It's a linked list.. you can add new nodes/tokens anywhere you want to; You just push the new tokens into the linked list and they will be printed - can even be a single token with an invalid type containing the full string of the node you want to insert.

This still does not handle dynamically created nodes. Having to manually create tokens if you want to create a new node is absurd and you're shifting codegen to the transformer. For a similar example on how this can be a nightmare see [jstransform](https://github.com/facebook/jstransform/blob/master/visitors/es6-class-visitors.js).

> ASTs are extremely valid for transpilers and static analysis, but not so much when it comes to lossless code transformation.

Lossless code transformation is also an extremely niche target which is why it wont be considered at the forefront of CST discussion.

---

This discussion is not very productive. It's clear that your goals are not the same as ours of creating an interoperable standard for all present tools so you're better of going and creating your own.

**Comment by mikesherov** [comment]
> Linked lists can't be serialized into JSON (how would you describe cross/circular references?) so the whole idea of "next" and "prev" tokens doesn't work out.

The Linked List can be represented via a dictionary of named nodes, with *trivial* client code for traversal

```
map = {
 1: { next: 2, previous: 0, value: 'b' },
 2: { next: 3, previous: 1, value: 'a' }
};
```

Client Code:
```
next = map[this.next];
prev = map[this.prev];
````

Fully serializable with linked list concept completely embedded. And this is also trivial to print.

> AST is not easily printable, see escodegen source code for an example. You need to infer a lot of things based on the tree structure since it doesn't contain all the tokens (if it doesn't contain all the parenthesis, colons, braces it is not a concrete syntax tree). Adding more properties to the tree will increase the codegen complexity even more...

Again, that's only if you consider AST that isn't linked by reference to the token linked list. Simple insertion algorithms can preserve both the AST portion and the linked token list. You retain the simple printing by eventually printing the token list, as you mentioned.

> if the AST contains range/loc info, it is not easy/cheap to mutate it (you need to update the loc info on all the nodes of the tree after each change).

You don't rely on range/loc. As long as AST has link to token list, the concrete src locations are discarded. You add comments in as first class citizens in terms of tokens, and you add `whiteSpacebefore` property. You have full lossless format without caring about locations.

> additive changes are not always backwards compatible since tools might rely on the current tree structure to work.

This is simply a case of not understanding what an additive change is. The current tree structure doesn't change, you only add to it, which can't break current tools.

**Comment by mikesherov** [comment]
The point here, of this whole thread, is that we need a set of principles and goals, and *then* people can propose strawmen, and *then* we can discuss downsides of each approach.

@jzaefferer @millermedeiros I understand that it may seem like I'm shutting this conversation down before people have had chance to prove themselves, but this are just wild discussions with no implementation to work off of. We need to agree on goals here.

**Comment by millermedeiros** [comment]
so OK, the goal is: "creating an interoperable standard for all present tools" - and I'll move on since that is not my goal. cheers.

**Comment by getify** [comment]
I just want to say something to those who seem from the tone here to be feeling a burden or exhaustion from this thread... I actually think this is precisely the kind of discussion that needs to be worked out *now*(ish) before we're too much further down the road. You may not like it, but I think it's a burden worth carrying.

No, I don't think we should be arguing about arrays vs linked lists, or what to name the properties. But the bigger issue, like whether a single tree can **and should** represent both abstract syntax and concrete syntax... *that* is precisely and entirely the point of why I'm here, and why I started this thread.

Clearly we don't all agree. Clearly ESTree is going to need to pick a stance on that, if it's going to handle CST at all.

I can understand and appreciate that CST is not seen by most here as an equally important goal to the goals of standardizing the ES6 parts of the existing AST form. I know there's this sense of needing to rush to get the rest of the tree done and then we can just come back to this topic later. I don't think that's entirely wise. I don't think these are orthogonal tasks. To some extent (not quite sure how much yet), they are related.

I think the strong level of disagreement about even high level goals here in and of itself argues that it's (perhaps unfortunately) some pain we need to take on in the short-term rather than putting off. This will be a lot harder and more painful if the entire ES6 AST is "finalized" and many tools are shipped and using it for 6 months, and only **then** we try to iron out these high level goal differences before we can get to implementation details.

I've seen this movie before. I know how it ends. It ends in good (and perhaps best?) designs being impossible because we didn't have enough foresight to paint ourselves into a corner. If we get to that point in the future, we failed on the spirit of why this thread was started. If we get to that point, we failed to address the bigger needs of the ecosystem in favor of only addressing a smaller and shorter term problem, and we will have yet again kicked this can down the road to make it someone else's problem.

**Comment by mikesherov** [comment]
@getify, this is getting ridiculous with the long diatribes. We're going to tackle this issue, we have a set of principles that will be a PR shortly, and we're going to move forward. The non starter is throwing out the AST as is. We're just not going to do that. Break back compat? Won't do that either.

I know you're anxious, but I'm telling you, we'll get there.

**Comment by mikesherov** [comment]
During the jQuerySF conference ( June 22-23 in San Fransisco California, http://jquerysf.com/ ) , the Esprima team will be having a face-to-face discussion for what a CST (Concrete Syntax Tree) might look like for Javascript, with hopes to standardize it as an add-on to the ESTree AST format.

We hope you can join us there in person, as your input to these discussions is extremely valuable, but if not, we'll make Skype available so that all interested parties can attend at least virtually. Please let me know if you can make it, and if not, if you plan on attending via Skype.

If you have any other concerns or questions, please feel free to let me know. I plan on also posting a public invite as an issue on the ESTree issue tracker. 

/cc @RReverser @millermedeiros @michaelficarra @getify @sebmck @estree/owners @dherman @nzakas @jzaefferer

**Comment by getify** [comment]
I for sure want to make every effort to be there. If at all possible, can this happen on the 23rd? I am already planning to fly into SF on the 23rd, and I can just do so early enough to attend this meeting without needing to change flight/hotel plans.

**Comment by RReverser** [comment]
I would definitely love to join, but will be talking at MediterráneaJS on these dates :( I'll try to join via Skype, but not sure about the schedule yet.

**Comment by getify** [comment]
Do we have any confirmation on when this F2F meeting will be? I haven't yet scheduled my flight into SF but I need to do so ASAP.

**Comment by mikesherov** [comment]
It's up to @ariya really. @ariya can you confirm the 23rd?

On Mon, Jun 8, 2015 at 4:52 PM, Kyle Simpson <notifications@github.com>
wrote:

> Do we have any confirmation on when this F2F meeting will be? I haven't
> yet scheduled my flight into SF but I need to do so ASAP.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/estree/estree/issues/41#issuecomment-110134696>.
>



-- 
Mike Sherov

**Comment by ariya** [comment]
@mikesherov  @getify  23rd works for me.

**Comment by getify** [comment]
OK, 23rd. What time and location? :)

**Comment by mikesherov** [comment]
I'm thinking either 2:30PM or 5:30PM... @ariya and @getify do those time work for you?

**Comment by mikesherov** [comment]
@getify The location would be at the jQuerySF conference.

**Comment by getify** [comment]
I would prefer 5:30pm if possible. The flights that get me in before 2:30pm are a bit more expensive. :)

**Comment by ariya** [comment]
@mikesherov @getify 5:30 pm works for me.

**Comment by mdevils** [comment]
Hello everyone. I'm not sure why I'm writing here in this thread, but @mikesherov wants discussion, so let it be.

The need for CST is a long story for JSCS (JavaScript Code Style checker, http://jscs.info):

— To check style for some situation we need to navigate: tree nodes -> token list and vice versa, including searching by type, range, starting token, ending token and so on.
— To check offset for most of the cases we included comments into the token list.
— To fix whitespace we hacked tokens with additional data (`whitespaceBefore`), so we could alter whitespace between certain tokens.

We hacked AST and token list to make things work together. But they do work fine to some limit: we can only alter simple things like whitespace, adding semicolons, additional commas. But requirements for fixing code style grow on daily basis. We need to be able to apply more serious changes: changing node order (for instance `1 === x` -> `x === 1`), converting function declarations to function expressions (`function f() {}` -> `var f = function() {}`) and so on... it's a long list.

In the beginning I was inspired by the idea of having new CST standard and parser implementation for that CST standard, but after a while I realised that in case of JSCS CST standard will not solve our problems and it even can harm us. I will try to explain why.

What we need in JSCS, big picture:

— Search for elements (by type, value) in the CST.
— Apply modifications (add/remove) to the CST node structure.
— Apply modifications (add/remove) to the CST token structure.
— Navigate through CST.
— Render final JavaScript source after all the modifications.

If we switch to some CST JSON format, then:

— How to rebuild CST search index on tree modification?
— How to update `range` and `loc` on every modification? Once you modify the tree, some ranges become invalid.
— How to make sure JSCS rule will not break syntax of the JS file?
— How to update all the navigation pointers (child -> parent, node -> tokens) on modifications?

So I came to a conclusion that for JSCS we don't need some kind of CST format, we need CST library which:

— Produces estree AST-compatible output.
— Adds tokens to the tree, including comment-tokens and whitespace-tokens.
— Builds and maintains search indexes automatically.
— Automatically maintains all the navigation links (i.e. `parentElement`, `childElements`, `nextToken`, `fistToken` and so on).
— Ensures CST is always valid from syntax point of view. Does not allow any modification which break syntax.
— Calculates `range` and `loc` on demand.
— Can render any part of the tree to JS code.

As a result I've created CST library with ES5 already implemented: https://github.com/mdevils/cst

With @hzoo's help we could integrate Babel/Acorn as parser (as Acorn is less buggy in terms of ranges then esprima).

This library provides AST-compatible tree which can be used in any tools which need to apply modifications to JS files preserving original style.


/all I'm open for the discussion, even though I'm not sure what to discuss.
@mikesherov can you please now ask questions?

**Comment by markelog** [comment]
I will try to enrich @mdevils answer a bit. 

Imagine HTML as JavaScript in this context, then AST + tokens would be HTML specification and CST would play role of DOM implementation.

In other words we think CST shouldn't be a format that defines static document but an API specification.

This is a whole different way of looking on the CST problem, but this is exactly what we need in jscs, it is pretty enlightening refactoring it using `cst` module and judging on what listed under "Why Concrete Syntax?" header, it should work just as well for other use-cases too.

**Comment by nzakas** [comment]
ESLint has been investigating how to augment the AST to allow fixing of errors as well. I really like the direction of @mdevils work.

**Comment by gibson042** [comment]
I find @mdevils work to be valuable, but don't think it replaces the need for a JSON-compatible CST extension of estree. The latter need not be much, but should at least capture the content and sequence of all lexical tokens (including comments, whitespace, and line terminators) (for e.g. #94 and babel/babel/issues/497) and connect AST nodes to them (for e.g. #90 and #92).

**Comment by getify** [comment]
I'd be very curious how two separate CST-enabled tools could "share" (aka "pipe") a CST data structure from one to the other? Think *nix CLI file descriptor pipes, not JS streams.

**Comment by gibson042** [comment]
There are several options; here's a nonexhaustive list:
* a top-level list of lexical tokens, referenced from syntax nodes by numerical index
* a top level hash of (autogenerated) identifiers to lexical tokens, referenced from syntax nodes by identifier
* a list of (lexical token|property name of associated subnode) associated with each syntax node, disjoint between them and complete over a `Program`.

Let me know if you'd like some code-level examples.

**Comment by getify** [comment]
It's not clear to me why the notion of CST as a format needs to change to CST as a library, since one of the base passing cases is that it must be serializable for persistence and transfer?

Each of those options you suggested for serialization sound reasonable at first glance, but why wouldn't we just decide on one of them as *the* CST standard, and then also release a CST helper library to do the extra tasks you're suggesting?

**Comment by gibson042** [comment]
> why wouldn't we just decide on one of them as _the_ CST standard, and then also release a CST helper library to do the extra tasks you're suggesting?

Are you confusing me with @mdevils? That's exactly what I'm suggesting.

**Comment by getify** [comment]
> Are you confusing me with @mdevils?

I am not confusing you. I wasn't really intentionally directing my questions/comments at any particular person, just asking in general. Sorry if it came off too directed at you.

**Comment by markelog** [comment]
I'm not sure why would you need to pipe anything, other tools could use `mdevils/cst` as dep, or implement the same API or use other tools with same API as a dep.

But let's say there is a use-case for that, lets say you need a static document from it,
then serialization would be the opposite process, since you get CST from AST + tokens, then serialization from CST to static document would the same AST + tokens.

**Comment by mikesherov** [comment]
Other tools having to use `mdevils/cst` is a non starter. The point is that not only javascript should be consuming the format.

**Comment by markelog** [comment]
> Other tools having to use mdevils/cst is a non starter

Not suggesting that, other tools have a choice to use it or choose another library... or another language with implemented <b>standart API</b>. 

Since we are talking here about specification right? Not to chose one library that we all like. Again, look at how different libraries on different language interact with XML document, through <b>standart</b> DOM API.

**Comment by gibson042** [comment]
Concrete proposal (pun intended) matching my [third suggestion](https://github.com/estree/estree/issues/41#issuecomment-120107587): an optional `Node#sourceElements` array allowing original source to be recovered by depth-first concatenation. This maintains JSON-friendly acyclicity and keeps all information both non-duplicated and node-local (for less complex transformation/manipulation). Source _rendering_ should prefer `sourceElements` when present, but verify that all child nodes are referenced in order (or alternatively insert them automatically along with necessary punctuators).

An example demonstrating representation of comments/whitespace and Directive Prologue edge cases:
```js
0,function/*anonymous*/( ) {
  // Not a Use Strict Directive
  "use\
 strict";

  // Not a directive at all
  ("use strict")
} // last character is a <LF>

```
yields the following:
```js
{
  "type": "Program",

  // new optional property: Node#sourceElements: [ SourceElement ]
  "sourceElements": [

    /* new interface
      Nonterminal <: SourceElement {
        reference: string (JSON Pointer relative to the parent AST node);
      }
    */
    // JSON Pointer could be replaced with a different suitably expressive language
    { "reference": "body/0" }
  ],
  "body": [
    {
      "type": "ExpressionStatement",
      "sourceElements": [
        { "reference": "expression" },

        /* new interface
          NonTokenTerminal <: SourceElement {
            element: string (WhiteSpace|LineTerminator|Comment*);
            value: string;
          }
        */
        { "element": "WhiteSpace", "value": " " },

        // Comment decomposition:
        // * `//`- or `/*`-prefixed head
        // * (possibly multi-line) body
        // * `*/`-suffixed tail, when applicable
        // Not strictly necessary, but may make manipulation & verification easier
        { "element": "CommentHead", "value": "//" },
        { "element": "CommentBody", "value": " last character is a <LF>" },
        { "element": "LineTerminator", "value": "\n" }
      ],
      "expression": {
        "type": "SequenceExpression",
        "sourceElements": [
          { "reference": "expressions/0" },

          /* new interface
            TokenTerminal <: SourceElement {
              element: string (Keyword|Identifier|Punctuator|Template*|*Literal);
              value: string;
            }
          */
          { "element": "Punctuator", "value": "," },
          { "reference": "expressions/1" }
        ],
        "expressions": [
          {
            "type": "Literal",
            "value": 0,

            // Note lack of `Literal#raw`—`sourceElements` subsumes such functionality
            "sourceElements": [
              { "element": "NumericLiteral", "value": "0" }
            ]
          },
          {
            "type": "FunctionExpression",
            "id": null,
            "params": [],
            "sourceElements": [
              { "element": "Keyword", "value": "function" },

              // Comments/whitespace can precede or follow any source element
              { "element": "CommentHead", "value": "/*" },
              { "element": "CommentBody", "value": "anonymous" },
              { "element": "CommentTail", "value": "*/" },
              { "element": "Punctuator", "value": "(" },
              { "element": "WhiteSpace", "value": " " },
              { "element": "Punctuator", "value": ")" },
              { "element": "WhiteSpace", "value": " " },
              { "reference": "body" }
            ],
            "body": {
              "type": "BlockStatement",
              "sourceElements": [
                { "element": "Punctuator", "value": "{" },
                { "reference": "body/0" },
                { "reference": "body/1" },
                { "element": "Punctuator", "value": "}" }
              ],
              "body": [
                {
                  "type": "ExpressionStatement",
                  "sourceElements": [
                    { "element": "LineTerminator", "value": "\n" },
                    { "element": "WhiteSpace", "value": "  " },
                    { "element": "CommentHead", "value": "//" },
                    { "element": "CommentBody", "value": " Not a Use Strict Directive" },
                    { "element": "LineTerminator", "value": "\n" },
                    { "element": "WhiteSpace", "value": "  " },
                    { "reference": "expression" },
                    { "element": "Punctuator", "value": ";" }
                  ],
                  "expression": {
                    "type": "Literal",

                    // `value` alone is not enough to detect the trickery
                    "value": "use strict",

                    // …but the corresponding StringLiteral is
                    "sourceElements": [
                      { "element": "StringLiteral", "value": "\"use\\\n strict\"" }
                    ]
                  }
                },
                {
                  "type": "ExpressionStatement",
                  "sourceElements": [
                    { "element": "LineTerminator", "value": "\n" },
                    { "element": "LineTerminator", "value": "\n" },
                    { "element": "WhiteSpace", "value": "  " },
                    { "element": "CommentHead", "value": "//" },
                    { "element": "CommentBody", "value": " Not a directive at all" },
                    { "element": "LineTerminator", "value": "\n" },
                    { "element": "WhiteSpace", "value": "  " },
                    { "reference": "expression" },
                    { "element": "LineTerminator", "value": "\n" }
                  ],
                  "expression": {
                    "type": "Literal",
                    "value": "use strict",
                    "sourceElements": [
                      // Even simple AST nodes can have multiple source elements
                      // Any non-";" Punctuator terminates a Directive Prologue,
                      // as would a non-ExpressionStatement or non-Literal or non-string
                      { "element": "Punctuator", "value": "(" },
                      { "element": "StringLiteral", "value": "\"use strict\"" },
                      { "element": "Punctuator", "value": ")" }
                    ]
                  }
                }
              ]
            }
          }
        ]
      }
    }
  ]
}
```

**Comment by RReverser** [comment]
@gibson042 Basically the same question as to any other proposed format - how are you going to keep AST and CST information in sync? Remember than any new structure should be backward-compatible and tools that handle AST only should be somehow resynced with CST data. Exact structure is not a problem (otherwise we would agree on some long time ago) but keeping data in sync is really a pain in... you know, head.

**Comment by gibson042** [comment]
> how are you going to keep AST and CST information in sync?

As I see it, that breaks down into more than one problem:
* **parsing**: Any node without `sourceElements` is still valid, so existing parsers are already compliant. And as long as `sourceElements` always includes at least the proper tokens (e.g., keywords and punctuators) and child-node references where it _does_ appear (which can be on a subset of the tree), it's essentially an AST fallback—although the whole point is obviously inclusion of whitespace, newlines, and comments as well.
* **node additions**:
  * Either omit `sourceElements` from the incoming node, or generate it with reasonable contents.
  * Either delete `sourceElements` on the parent, or add a reference to the incoming node.
* **node moves**:
  * Either delete `sourceElements` on the old parent, or remove the reference to the node and update all following references (_side note: this suggests that a selection language like `<property>#next` might be superior to [JSON Pointer's](https://tools.ietf.org/html/rfc6901#section-3) `<property>/<N>`_).
  * Either delete `sourceElements` on the new parent, or add a reference to the incoming node.
  * If desired, update whitespace/newlines associated with the migrating node.
* **literal manipulation**: Either delete `sourceElements`, or update its `*Literal`.
* **other node manipulation**: It's probably best to delete `sourceElements`, but obviously possible to sync it manually (e.g., by reusing comments and reusing/replacing whitespace and newlines).
* **rendering**: Where `sourceElements` is defined, some sanity checking as described [above](https://github.com/estree/estree/issues/41#issuecomment-131636382) would be prudent.

> any new structure should be backward-compatible

:white_check_mark: :+1:

> tools that handle AST only should be somehow resynced with CST data

`delete node.sourceElements` and sometimes `delete parent.sourceElements`, or update them along with other changes. Rendering operations can also be smart enough to detect mismatches, and in such cases might ignore `sourceElements` and generate a warning. Addition of these not-originally-planned elements to syntax trees certainly introduces complexity, but this approach attempts to minimize it by allowing for graceful fallback and gradual adoption.

**Comment by hzoo** [comment]
Just an update that we've finished up es6 nodes in https://github.com/cst/cst (tracking proposals [here](https://github.com/cst/cst/issues/39)) and planning on jscs integration soon to test it out

**Comment by donaldpipowitch** [comment]
Nice! Thank you for the update. Didn't know about `cst` before.

**Comment by gibson042** [comment]
@hzoo How compatible is your internal representation with [my proposal](#issuecomment-131636382)? Or more to the point, how much effort would it take to input and output documents in such an updated ESTree format?

**Comment by markelog** [comment]
@gibson042 CST module works with AST and tokens only, so if we change format, we would need to change everything

**Comment by nzakas** [comment]
Why are comments separated into parts? That seems like a good way to create individual syntax accidentally (alter CommentHead to change from block to line, then forget to remove CommentTail).

**Comment by mikesherov** [comment]
Seeing as CST is WIP on JSCS, ideal is to have consensus formed here and documented. Lots of smart folk in this repo.

Ideally, even when CST is "finished" in JSCS, we should be striving to ultimately make that work useless by moving that work into parsers and generators anyway.

Nice work @hzoo and @markelog on getting a real POC out there!

**Comment by gibson042** [comment]
> Why are comments separated into parts? That seems like a good way to create individual syntax accidentally (alter CommentHead to change from block to line, then forget to remove CommentTail).

@nzakas This is probably better answered on #107, but I would point out the analogous risk of modifying an atomic `Comment` in such a way that either the head or a necessary tail is removed—making the head and tail discrete is just an effort to capture the structure. But I'm not married to the idea, and would be happy to change it.

**Comment by forivall** [comment]
My WIP names have been `BlockCommentStart`, `BlockCommentBody`, `BlockCommentEnd`, `LineCommentStart` and `LineCommentBody`. But I'm not married to these specific names either. However, I will need to make the head and body discrete, and it would be nice for that to be in this spec.

Or, the `CommentTail` of Line comments could have either a value of `""` or `"\n"`

**Comment by nzakas** [comment]
Some other use cases I think it's worthwhile to think through:

1. Given a node, determine if it is surrounded by parentheses. 
2. Insert a new argument between two existing arguments of a CallExpression.
3. Comment out a function.
4. Combine multiple `var` statements into a single `var` statement.

**Comment by gibson042** [comment]
@nzakas: Answered [on the PR](https://github.com/estree/estree/pull/107#issuecomment-154557885).

**Comment by mdevils** [comment]
Hello guys. I would say `cst` package is reaching maturity level: https://github.com/cst/cst

ES6 is completely implemented along with JSX extensions. Plus we managed to implement `Scopes` in a live tree (scopes are updating while CST tree is being changed). We have almost finished integrating CST into JSCS.

**Comment by DanielSWolf** [comment]
Has there been any progress on this? I'm using Babel to migrate some in-house code. It's working fine, but creates all kinds of whitespace changes.

**Comment by danez** [comment]
@DanielSWolf You might be better off with using https://github.com/square/babel-codemod. Babel by default uses `babel-generator` which is meant to output generated sourcecode and not necessarily readable code for performance reasons. `babel-codemod` uses `recast` to print the AST which tries to preserve the original code as good as possible.

**Comment by DanielSWolf** [comment]
Thanks @danez, I didn't know about that one! The only downside seems to be that it's not meant to be used as a library, but as a CLI. But that shouldn't be much of a problem.

**Comment by KvanTTT** [comment]
Another name is the full-fidelity tree. See the similar [issue](https://github.com/jquery/esprima/issues/1952) from esprima project.

**Comment by conartist6** [comment]
Hey guys, I read through this whoooole thread and build a prototype of a library that handles everything: https://github.com/conartist6/cst-traverse. It's similar in scope to `cst`, but where that library uses its own parser and attempts to prevent invalid changes to the resulting tree, my library can work with any parser and embraces the idea of a malformed tree. It is specifically optimized to ensure that tools that are completely unaware of concrete syntax can do anything they like with the tree and as much of the original concrete syntax as is still valid for the new AST will be used.

**Comment by conartist6** [comment]
@gibson042 It's worth nothing that this implementation is pretty much exactly what you proposed.

**Comment by gibson042** [comment]
@conartist6 That's great news!

**Comment by conartist6** [comment]
I've gotten a lot done, and also shifted my approach. I said in my last post that I intended to make sense of malformed trees. I no longer intend to do this, as it is not a fundamentally scalable approach. Instead my focus is on validation: thoroughly defining which tree structures are valid.

Here's what it currently looks like to validate/retokenize this simple code:
```js
[1 ]
```

<img width="404" alt="cst-tokens-run-2x" src="https://user-images.githubusercontent.com/540777/197350819-f1e179fc-e32d-4c3f-8129-4a65ba18967d.png">

Those of you with sharp eyes will notice that my test grammar is not matching comments, but I already have a more complete implementation that matches comments and also preserves all valid escapes in strings and identifiers.

The next problems I intend to tackle are defining valid trees with holes in them, and building grammar-aware APIs for mutating and composing trees.

**Comment by conartist6** [comment]
Something else really cool to note about the example above: notice these two lines:
```
branch (at Array.cstTokens[1])
accept (at Digit.cstTokens[0])
```

What you are seeing is evidence of my hoisting algorithm. I have defined the canonical representation of any tree to be the one in which all mergeable tokens are merged and all hoistable tokens are hoisted. Whitespace tokens are both mergeable and hoistable, so canonical trees will never have whitespace at the beginning or end of `node.cstTokens` except at the root node.

I'm particularly proud of the way my engine can handle both non-canonical trees and non-canonical grammars though. The benefits of full support for non-canonical trees and grammars cannot be overstated!

Non-canonical trees allow us to push certain spaces and comments down into the tree. For example if you have a suppression comment like `prettier-ignore` that needs to stay stuck to a particular node for its meaning to be preserved, you would simply push that comment token down into that node. Now you have a non-canonical tree, but you can still traverse it as if it were a canonical tree, and more importantly if the node moves in the tree the comment moves with it. A single pass can be made over the tree to push down whitespace and comments according to user-defined logic, after which nodes can be reordered freely without worrying about changing the meaning of the code.

Non-canonical grammars are those which those which sometimes match hoistable tokens lower in the tree than they could. In fact the test grammar shown in the example does this. Each node in the [test grammar](https://github.com/js-cst-tokens/cst-tokens/blob/9058d055a10c33546b20e91d74f1314e23f2d762/test/grammars/digit-array.mjs#L130-L131) is responsible for matching leading space but not trailing space. Even so, the trees produced are always canonical. Ultimately this allows me to handle whitespace in a higher order grammar: My javascript grammar fully automates the process of defining where whitespace and comments are allowed and required. For example whitespace is allowed between a `Punctuator` and an `Identifier`, but it is required between an `Identifier` and an `Identifier`. This was initially a novelty, but it turns out to be requisite to correctly understand some more complex cases, particularly those at node boundaries. For example consider `for (const [i]of bar)` and `for (const i of bar)`. Is space required between `forOfExpr.initializer` and `of`? As you can see, the answer depends on what is contained in `forOfExpr.initializer`.

**Comment by conartist6** [comment]
Another interesting thought about the data structure: when I was intending to do best-effort munging of invalid trees it was important that references be a bit ambiguous.

In my design you might have:
```js
{
  type: 'Array',
  elements: [/*...*/],
  cstTokens: [
    /*...*/,
    { type: 'Reference', value: 'elements' },
    { type: 'Punctuator', value: ',' },
    { type: 'Reference', value: 'elements' }
    /*...*/,
  ]
}
```

A resolver is used so that the first `elements` token references `elements[0]` and the second references `elements[1]`. I took the inspiration from URL parameter encoding actually.

At the moment though there's no longer any real need to do this that I know of, so I could do something like what @gibson042 suggested and use { type: 'Reference', value: 'elements/0' }.

**Comment by conartist6** [comment]
Ah, now that I think about it resolvers are crucial. Without them adding an element to the beginning of an array might involve editing every other reference token to change its index!

**Comment by conartist6** [comment]
I've also made some innovations in how to break down tokens. Moving tokens into a new well-known location in the syntax tree would provide a one-time opportunity to make breaking changes to the way tokens are defined, so I've put quite a bit of consideration into how the tokenization could better support the kinds of tasks that tokens are likely to be used for. Many of these ideas originated in this thread!

- Pseudo keywords like import/export `as` changed from `Identifier` to `Keyword`.
  - Ensures `Identifier` is reserved for symbol definitions and references
- All `\n` characters (even those inside literals) are required to be contained within `LineBreak` tokens!
  - LineBreak tokens still have a value recording whether they were '\r', '\r\n` or `\n`
  - Makes it easy to  safely (or unsafely) change line endings as needed
  - Should make it easier to integrate CSTs with line-driven tools
- Block comments are: `BlockCommentStart`, `Comment`, `BlockCommentEnd`
- Line comments are: `LineCommentStart`, `Comment`, `LineBreak` (or EOF)
- Strings are: `StringStart`, `String`, `StringEnd`
- A possible template string is: `StringStart`, `LineBreak`, `InterpolateStart`, `Reference`, `InterpolateEnd`, `String`, `StringEnd`
- `Punctuator` is joined by `LeftPunctuator` and `RightPunctuator`
  - Supports language-agnostic balanced brace matching
    -  Relevant features include block folding and code search
- `yield*` is now a single `Keyword`

I think a general theme here is ensuring that the token structure tells you as much as it can about the nature of the language, where the current token structure often expects you to know things about the language to make use of it.

**Comment by conartist6** [comment]
A couple fun things about the system I've shared:

- All the processing is time-structured (streaming). Each line of the output is a "token" in the parser's output stream. You can see the complete log of this run in [this gist](https://gist.github.com/conartist6/3b6b485db3d4e38f6085d7d0abdbdb43), with external emits marked as `>>>`.
- Grammars require no compilation, are extensible through class extension, and have no dependency on the core.
- You can always get the original document back by printing the contents of the string-quoted tokens. Escapes get a dual representation in the document that looks like this `!'&amp;' :'&'`, ensuring that either a "cooked" or a "raw" form of content can be constructed as needed without additional language-specific knowledge.
- JSON is defined in an "atrivial" grammar which is wrapped in a higher-order trivia grammar. Because the grammar yields pattern-matching calls to the VM executing it, a higher order grammar is actually "closer to the metal", a fact that we can exploit to inject trivia-matching! Basically this ensures that you can extend the JSON language (if you needed to) then re-apply the trivia enhancers to create your whitespace-only and comment-allowing variants of the language.

**Comment by conartist6** [comment]
The solution would blur if not completely eliminate the distinctions between ESLint, Prettier, Babel, and JSCodeshift, while ensuring that the complete suite of tools can be used when working with code written in any language.

**Comment by conartist6** [comment]
Just shipped a new release, and finally it has a JS parser that works pretty darn well.
```bash
npx -p @bablr/cli -p @bablr/language-en-es3 bablr -l @bablr/language-en-es3 -p Program -s << EOF
var a = 2,
    b = a + (2 * 3);
EOF
```
